Adds confirmation when changing segments definition

sgiehl opened the pull request:
Saving a changed segment definition will now trigger a confirmation:

![image](https://cloud.githubusercontent.com/assets/1579355/15690988/8bd4eb14-2786-11e6-92d0-2c429fd55359.png)

Maybe there's a better text to show. Suggestions welcome

fixes #9984

sgiehl committed:
refs #9984 - adds confirmation when changing segments definition

sgiehl committed:
refs #9984 - adjust UI test to automatically confirm confirmation

adaqus commented:
Looks ok :+1: 

Though maybe it would be better to change 'segment-change-confirm' to 'segment-definition-change-confirm' and 'ChangingSegmentConfirmation' to 'ChangingSegmentDefinitionConfirmation' to point out this functionality concerns segment definition change?

sgiehl committed:
better naming

sgiehl commented:
@adaqus thanks for the comment. I've changed the name/id as you suggested.

@mattab is the message for the confirmation fine or should it be changed? If not, I would merge this PR.

mattab commented:
Review:

* let's add a UI test for this use case, so it won't regress
* change text to: `You're about to change the segment definition. Your analytics reports for this new segment won't be available until the reports are re-processed. It may take a few hours to reprocess your reports. Proceed anyway?`

sgiehl commented:
Ok. I've changed the text and added a ui test:
http://builds-artifacts.piwik.org/piwik/piwik/master/19275/SegmentSelectorEditorTest_update_confirmation.png

mattab commented:
just realised we can make things a bit more clear and reduce some possible confusion from the wording:

* `if( ! Rules::isRequestAuthorizedToArchive() || globalConfig.General.browser_archiving_disabled_enforce ) ` then print the message `Your analytics reports for this new segment won't be available until the reports are re-processed. It may take a few hours for reports data to show for this segment.`. 
* Otherwise (ie. case where reports will be processed in the browser), maybe write `Your analytics reports for this new segment will be re-processed the next time you request them, which may take some time. `

(the if statement logic reproduces some sub logic of `Rules::isArchivingDisabledFor()` and can be added as a method there)

After this change LGTM and feel free to merge!

sgiehl committed:
use better message depending of the availability of archiving triggered by browser

sgiehl commented:
@mattab I've tried to implement your suggestion. Not sure if I got that correct. Would you please have a look at https://github.com/piwik/piwik/pull/10199/commits/cd05517bc0faced2ccf013902cb33b258da365bd
If it's correct, feel free to merge,

Note: We need to update the UI test screenshot afterwards

mattab commented:
Review:

* just saw again that our segments can individually be set to be pre-processed only, so let's make the message variant depending on the `segment.auto_archive` 
* after clicking Proceed Anyway -> No, then editing the segment and clicking Save again, nothing happens (no popup, no save).
* If users often edit segments, the message may get a bit too much... I'm wondering whether:
** we could hide the message completely when browser archiving is available for this segment?
** or maybe we could add a check box `[ ] Hide this notice in the future` that would be saved in the user preference?

mattab commented on plugins/SegmentEditor/lang/en.json:
```diff
@@ -6,6 +6,8 @@
         "AreYouSureDeleteSegment": "Are you sure you want to delete this segment?",
         "AutoArchivePreProcessed": "segmented reports are pre-processed (faster, requires cron)",
         "AutoArchiveRealTime": "segmented reports are processed in real time",
+        "ChangingSegmentDefinitionConfirmationNotProcessedOnRequest": "You're about to change the segment definition. Your analytics reports for this new segment won't be available until the reports are re-processed. It may take a few hours for reports data to show for this segment. Proceed anyway?",
```
maybe refactor in separate translation strings the `You're about to change the segment definition.` and `Proceed anyway?` ?

sgiehl replied:
Sure we could do that. But maybe it's better to have complete messages as one string. That prevents from having only parts of it translated, which would look very ugly in the UI I think.

sgiehl commented:
> just saw again that our segments can individually be set to be pre-processed only, so let's make the message variant depending on the `segment.auto_archive` as well

I'm a bit confused now. What will happen when the segment setting and the config setting differ? Will a segment not be processed at all if I choose real time processing for the segment, but have a config setting enforcing not to do it? (or vice versa)

mattab commented:
> What will happen when the segment setting and the config setting differ? Will a segment not be processed at all if I choose real time processing for the segment, but have a config setting enforcing not to do it? (or vice versa)

If a segment is set to real time processing, but browser archiving (real time processing) is disabled in the UI, and enforced via `General.browser_archiving_disabled_enforce=1` in config, then the segment will be processed in the next `core:archive` run (via crontab). 


FYI: if a segment is set to "real time processing" but the browser archiving is disabled and enforced, then the data for the segment cannot be processed, and we show this popup:

![segment not applied](https://cloud.githubusercontent.com/assets/466765/16822239/1fa34bf2-49af-11e6-8a82-c42bb7ae5a30.png)

sgiehl committed:
add option to hide segment definition change confirm

sgiehl commented:
New attempt ;) Confirmation should now also work as expected (more than once). Message now also depends on segments auto-archive option, but will only be displayed if browser archiving is not available. 
I've also implemented an checkbox to hide the message in the future.

mattab commented:
>      If a segment is set to real time processing, but browser archiving (real time processing) is disabled in the UI, and enforced via General.browser_archiving_disabled_enforce=1 in config, then the segment will be processed in the next core:archive run (via crontab). 

fyi: after double checking what I wrote yesterday I realised it wasn't actually working this way. Created bug report: https://github.com/piwik/piwik/issues/10307 and PR: https://github.com/piwik/piwik/pull/10309

sgiehl commented:
Ok I'll have a look at your PR. What about this one? Did you have another look? Can we merge?
